gluon-config-mode-geo-location-with-map: integrate a location picker map

2tata opened the pull request:
This MR is a split of #1211 
Its add an osm map into the config-mode (in case the clients computer has an internet connection).

![screenshot_2018-07-18_07-00-21](https://user-images.githubusercontent.com/3101274/42860813-98a787b6-8a58-11e8-8068-ee304cfe9d9c.png)

Issue #58 

- [x] export zoom level to site.conf.

- [x] check translation.

Signed-off-by: Jan-Tarek Butt <tarek@ring0.de>

CodeFetch commented:
Thank you for your work. We will surely have users entering more accurate coordinates with this.
I'd prefer checking for internet access using JavaScript instead of showing a broken map.
Is there a better way for translating such long sentences? 
The package description should say „manually“.

2tata commented:
On 07/18/18 22:47, Vincent Wiemann wrote:
> Thank you for your work. We will surely have users entering more accurate coordinates with this.
> I'd prefer checking for internet access using JavaScript instead of showing a broken map.

The map will only shown if the clients computer has an internet connection.
Otherwise it will not be displayed. Means it will lock like the old solution.

> Is there a better way for translating such long sentences?
> The package description should say „manually“.

There are multiple packages in one Makefile. I was discussing with @NeoRaider about various solutions and after few try`s I was landed on this solution. Which has less redundancy code and  backward compatibility for people who would preferring the current solution without map.

2tata commented:
@CodeFetch Now the map is only visible on internet connection.

rotanid commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,62 @@ PKG_VERSION:=1
 
 define Package/gluon-config-mode-geo-location
-  TITLE:=Set geographic location of a node
+  TITLE:=Set geographic location of a node manualy and share it.
```
"manually" with two "l"

rotanid commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,62 @@ PKG_VERSION:=1
 endef
 
+define Package/gluon-config-mode-geo-location-with-map
+  TITLE:=Set geographic location of a node manualy and share it, optionally show a map if internet available.
```
"manually" with two "l"
"if internet is available" (adding the "is")

rotanid commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+"Koordinaten. Hier hast Du die Möglichkeit, "
+"die Koordinaten händisch zu hinterlegen."
+#ifdef WITHMAP
+"Wenn dein Computer mit dem du den Router einrichtes am Internet angeschlossen ist, "
```
"den Router einrichtest eine Internetverbindung hat,"

rotanid commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+"die Koordinaten händisch zu hinterlegen."
+#ifdef WITHMAP
+"Wenn dein Computer mit dem du den Router einrichtes am Internet angeschlossen ist, "
+"hast du die Möglichkeit auf der unten angezeigten Karte an die stelle zu klicken wo "
```
"Stelle" Großschreibung
"zu klicken wo" -> "zu klicken, an der"

rotanid commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+#ifdef WITHMAP
+"Wenn dein Computer mit dem du den Router einrichtes am Internet angeschlossen ist, "
+"hast du die Möglichkeit auf der unten angezeigten Karte an die stelle zu klicken wo "
+"der Router Stehen wird."
```
"stehen" -> Kleinschreibung

rotanid commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+"hast du die Möglichkeit auf der unten angezeigten Karte an die stelle zu klicken wo "
+"der Router Stehen wird."
+#endif
+"Bitte berücksichtige das, das setzen einer Position die Netzwerk Qualität verbessern kann."
```
"berücksichtige, dass das Setzen"

rubo77 commented:
Do I still have to include this block in my `site.conf` then?

    config_mode = {
                geo_location = {
                        map_lon = 52.951947558,
                        map_lat = 7.844238281,
                        olurl = 'http://osm.ffnw.de/.static/ol/OpenLayers.js'
                        show_altitude = true,
                },
    },



I built your branch https://github.com/2tata/gluon/ with GLUON_BRANCH: gluon-config-mode-geo-location, but there is no map visible (without the extra block in site.conf)

Do I have to include more than just the package `gluon-config-mode-geo-location` in my `site.mk`?

I think you should add a site in the docs/ section or a README to the package to explain what there is to do to include this package.

2tata commented:
On 07/20/18 17:00, Ruben Barkow wrote:
> Do I still have to include this block in my `site.conf` then?
> 
>      config_mode = {
>                  geo_location = {
>                          map_lon = 52.951947558,
>                          map_lat = 7.844238281,
>                          olurl = 'http://osm.ffnw.de/.static/ol/OpenLayers.js'
>                          show_altitude = true,
>                  },
>      },

This site.conf section is just optional,

Its replace the default position of the map centre (default: lat:0.0 & lon:0.0) 
and the openlayers URL, if some one would like to insert its one map (default: 
"http://dev.openlayers.org/OpenLayers.js")

> I built your branch https://github.com/2tata/gluon/ with GLUON_BRANCH: gluon-config-mode-geo-location, but there is no map visible (without the extra block in site.conf)
> 
> Do I have to include more than just the package `gluon-config-mode-geo-location` in my `site.mk`?

The package `gluon-config-mode-geo-location` will be still available. If you 
want to have the map inside you need to include 
`gluon-config-mode-geo-location-with-map`.

> I think you should add a site in the docs/ section or a README to the package to explain what there is to do to include this package.

I will add some doc during the day.

cheers
Tarek

rubo77 commented:
> If you want to have the map inside you need to include
`gluon-config-mode-geo-location-with-map`.

only that one? or do I have to add `gluon-config-mode-geo-location` too?

2tata commented:
On 07/20/18 18:13, Ruben Barkow wrote:
>> If you want to have the map inside you need to include
> `gluon-config-mode-geo-location-with-map`.
> 
> only that one? or do I have to add `gluon-config-mode-geo-location` too?

No, both packages are in conflict only one of both is possible.

cheers
Tarek

2tata commented:
Now, there is some doc available.

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+gluon-config-mode-geo-location-with-map
+---------------------------------------
+
+The package ``gluon-config-mode-geo-location-with-map`` will additionally and a
```
"and" ? do you mean "add" ?

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+---------------------------------------
+
+The package ``gluon-config-mode-geo-location-with-map`` will additionally and a
+map where users can pick a position. This map will only shown if the users
```
"only shown" -> "only be shown"

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+
+The package ``gluon-config-mode-geo-location-with-map`` will additionally and a
+map where users can pick a position. This map will only shown if the users
+computer will have an internet connection.
```
"will have" -> "has"

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+site.conf
+^^^^^^^^^
+
+This option is valid for both typ of package.
```
"typ of package" -> "package variants"  <-- not sure though about the meaning

2tata replied:
Your version sounds better.

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+  - represents the default latitude center of the location picker map.
+  - defaults to ``0.0``
+
+The above 2 options will usually shown on a factory flashed Router. If a node
```
"shown" -> "be shown"

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+  - defaults to ``0.0``
+
+The above 2 options will usually shown on a factory flashed Router. If a node
+is reenter the config mode the maps center will be on the last defined
```
"is reenter" -> "re-enters" --> not 100% sure though
"the maps" -> "the map's" (genitiv s in english)
"will be" -> "is" (not sure though)

rotanid commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,136 @@
+        local pkg_i18n = i18n 'gluon-config-mode-geo-location'
+        local site_i18n = i18n 'gluon-site'
+
+        local site = require 'gluon.site'
```
check indentation in this file please, this line and the lines before have a different indentation

2tata replied:
Oh yes, indeed ... wrong configuration in vim..

rotanid commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,136 @@
+
+#endif
+
+        -- show altitude if site.json retruns true or nil and uci an value. Dont show if site.json retruns false or uci nil!
```
one space too mach indentation.
"retruns" -> "returns"
"Dont" -> "Don't"

rotanid commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,136 @@
+
+#ifdef WITHMAP
+
+		'If your PC is connected to the internet you can also click on the map displayed below. ' ..
```
"can" -> "may"

rotanid commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+"If you want the location of your node to be displayed on the map, you can "
+"enter its coordinates here. "
+#ifdef WITHMAP
+"If your PC is connected to the internet you can also click on the map displayed below. "
```
"can" -> "may" (same as before)

rotanid commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+#endif
+"Please keep in mind setting a location can also enhance the network quality."
+msgstr ""
+"Um Deinen Router auf der Karte anzeigen zu können, benötigen wir seine "
```
please search for "Deinen" and "Du" and make it all lowercase (if it's not the beginning of the sentence)
The uppercase version isn't valid anymore and hasn't been for many years - afaik.

rotanid commented on package/gluon-config-mode-geo-location/src/fr.po.in:
```diff
@@ -0,0 +1,51 @@
+"Plural-Forms: nplurals=2; plural=(n != 1);\n"
+
+msgid ""
+"If you want the location of your node to be displayed on the map, you can "
```
"can" -> "may"  (same as above)

2tata replied:
I think you just mean the second "you can"

rubo77 replied:
"If you want the location of your node to be displayed on the map, you may "

2tata replied:
I think in the first sentence should be you can and in the second you may to have a stronger suggestion on the map.

rotanid commented on package/gluon-config-mode-geo-location/src/fr.po.in:
```diff
@@ -0,0 +1,51 @@
+"If you want the location of your node to be displayed on the map, you can "
+"enter its coordinates here. "
+#ifdef WITHMAP
+"If your PC is connected to the internet you can also click on the map displayed below. "
```
"can" -> "may" (same as above)

rotanid commented on package/gluon-config-mode-geo-location/src/gluon-config-mode-geo-location.pot.in:
```diff
@@ -0,0 +1,32 @@
+msgstr "Content-Type: text/plain; charset=UTF-8"
+
+msgid ""
+"If you want the location of your node to be displayed on the map, you can "
```
"can" -> "may" (same as above)

2tata replied:
Same as above?

rotanid commented on package/gluon-config-mode-geo-location/src/gluon-config-mode-geo-location.pot.in:
```diff
@@ -0,0 +1,32 @@
+"If you want the location of your node to be displayed on the map, you can "
+"enter its coordinates here. "
+#ifdef WITHMAP
+"If your PC is connected to the internet you can also click on the map displayed below. "
```
"can" -> "may" (same as above)

rotanid commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -35,8 +35,8 @@ config_mode.geo_location.map_lat \: optional
-The above 2 options will usually shown on a factory flashed Router. If a node
-is reenter the config mode the maps center will be on the last defined
+The above 2 options will usually be shown on a factory flashed Router. If a node
+is reentered the config mode the map's center will be on the last defined
```
"a node is reentered" is still wrong, i'm not sure about the ideal way to say it... @mweinelt ?

mweinelt commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+  config_mode = {
+            geo_location = {
+                    map_lon = 52.951947558,
+                    map_lat = 7.844238281,
```
Expose zoom level and maybe tiles basepath? If we have this many options, create a subsection.

```
map = {
  lat = 52.951947558,
  lon = 7.844238281,
  zoom = 19,
},
```

2tata replied:
Not sure here.. This will need a lot more change also in the gluon-web-osm #1483 . I will need a bit time for that. But the exporting zoomlevel is a good point. I think the subsection is not necessary here because this options will be still related to the geo_location subsection or do you mean a subsection in the geo_location subsection?

mweinelt commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+            geo_location = {
+                    map_lon = 52.951947558,
+                    map_lat = 7.844238281,
+                    olurl = 'http://osm.ffnw.de/.static/ol/OpenLayers.js',
```
OpenLayers is the javascript used to create the map, so this option only decides where to pull the  file from, correct?

`olurl` is an abbrevation that is probably not easily understood, how about `openlayers_js_url` instead?

mweinelt commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,62 @@ PKG_VERSION:=1
 
 define Package/gluon-config-mode-geo-location
-  TITLE:=Set geographic location of a node
+  TITLE:=Set geographic location of a node manually and share it.
```
I'd say drop `manually`, because that is likely the default expectation.

mweinelt commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,62 @@ PKG_VERSION:=1
 endef
 
+define Package/gluon-config-mode-geo-location-with-map
+  TITLE:=Set geographic location of a node manually and share it, optionally show a map if internet is available.
```
-manually

> share it, **additionally** show a map if internet is available.

mweinelt commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,62 @@ PKG_VERSION:=1
+
+define Package/gluon-config-mode-geo-location/postinst
+#!/bin/sh
+$(call GluonCheckSite,$(PKG_BUILD_DIR)/gluon-config-mode-geo-location/check_site.lua)
```
Indentation missing, or is the shebang preventing that?

rubo77 replied:
I think the shebang prevents indentation here, like in `/gluon/package/gluon-config-mode-geo-location/Makefile`

2tata replied:
This seems necessary. I just took that from an older gluon make file. Also it is defined similar in gluon.mk. I think it has something to do wether it will interpreted as Makefile command (with Indentation) or not.

mweinelt commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,62 @@ PKG_VERSION:=1
+
+define Package/gluon-config-mode-geo-location-with-map/postinst
+#!/bin/sh
+$(call GluonCheckSite,$(PKG_BUILD_DIR)/gluon-config-mode-geo-location-with-map/check_site.lua)
```
see above

rubo77 replied:
this is obsolete then

mweinelt commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,136 @@
+#ifdef WITHMAP
+
+	local function show_lon()
+	  	if uci_longitude ~= nil then
```
tabs instead of spaces

2tata replied:
applied

mweinelt commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,136 @@
+	end
+
+	local function show_lat()
+	  	if uci_latitude ~= nil then
```
tabs instead of spaces

2tata replied:
applied

mweinelt commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,136 @@
+
+	local s = form:section(Section, nil, text)
+
+
```
remove newline

mweinelt commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+#ifdef WITHMAP
+"If your PC is connected to the internet you may also click on the map displayed below. "
+#endif
+"Please keep in mind setting a location can also enhance the network quality."
```
that sounds pretty far fetched

rubo77 replied:
maybe: "Please keep in mind setting a location gives you a better opportunity to get contact with mesh-partners that live nearby."

2tata replied:
I keep this quite abstract, because it can enhance the network quality in the sense of a good public map where users can find the next node or some kind of that. Another case can be for the hoodselector which can determinate the nodes hood base on its location.

mweinelt commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+"Please keep in mind setting a location can also enhance the network quality."
+msgstr ""
+"Um deinen Router auf der Karte anzeigen zu können, benötigen wir seine "
+"Koordinaten. Hier hast du die Möglichkeit, "
```
Position

2tata replied:
I think it should be "Koordinaten" which describes the nodes "Position"

mweinelt replied:
Talking about a position and then asking for coordinates for that position sounds more wholesome to me, that's why I recommended it that way.

mweinelt commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+msgstr ""
+"Um deinen Router auf der Karte anzeigen zu können, benötigen wir seine "
+"Koordinaten. Hier hast du die Möglichkeit, "
+"die Koordinaten händisch zu hinterlegen."
```
-händisch

2tata replied:
I think its better not to change it.

rubo77 replied:
"die Koordinaten von Hand einzutragen."

2tata replied:
@rubo77 ok, I'll apply your change.

mweinelt commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+#ifdef WITHMAP
+"Wenn dein Computer mit dem du den Router einrichtest eine Internetverbindung hat, "
+"hast du die Möglichkeit auf der unten angezeigten Karte an die Stelle zu klicken, an der "
+"der Router stehen wird."
```
hast du die Möglichkeit die Position des Routers auf der Karte auszuwählen.

mweinelt commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,54 @@
+msgstr "Deaktiviert"
+
+msgid "Share your location to see your router on the map"
+msgstr "Position für die Karte freigeben"
```
Odd,  but a map might not be the only application receiving that data, so maybe generalize?

2tata replied:
I didn't change anything there. This is from the master fork.

rubo77 replied:
so this should be an issue in the main gluon repository then

rubo77 commented:
How do you build this package in your testing environment? It seems like there are more packages needed at once, that are not present in the same repository at the moment:

```
'package/feeds/gluon_base/gluon-config-mode-geo-location/Makefile' has a dependency on 'gluon-web-osm', which does not exist
WARNING: Makefile 'package/feeds/gluon_base/gluon-config-mode-geo-location/Makefile' has a dependency on 'gluon-geolocator', which does not exist
WARNING: Makefile 'package/feeds/gluon_base/gluon-config-mode-geo-location/Makefile' has a dependency on 'gluon-geolocator', which does not exist
```

Te dependency to `gluon-web-osm` and `gluon-geolocator` should be noted in the readme

please add an example to the readme, which packages you have to include in your `site.mk`

rubo77 commented:
No offense, but also maybe you should use those tools to improve the initial english translations: 

- https://www.deepl.com/translator
- https://www.grammarly.com/

alternatively I would suggest to just add german texts for a start and let it translate by someone. So you don't have to hazzle with enlish and can concentrate on the content of your brilliant work

2tata commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -18,12 +18,12 @@ msgid ""
 "enter its coordinates here. "
 #ifdef WITHMAP
-"If your PC is connected to the internet you can also click on the map displayed below. "
+"If your PC is connected to the internet you may also click on the map displayed below. "
```
Not sure here. I think both variants are ok.

mweinelt replied:
selecting a position on the map is more specific than just clicking the map

2tata commented:
On 07/21/18 04:44, Ruben Barkow wrote:
> How do you build this package in your testing environment? It seems like there are more packages needed at once, that are not present in the same repository at the moment:
> 
> ```
> 'package/feeds/gluon_base/gluon-config-mode-geo-location/Makefile' has a dependency on 'gluon-web-osm', which does not exist
> WARNING: Makefile 'package/feeds/gluon_base/gluon-config-mode-geo-location/Makefile' has a dependency on 'gluon-geolocator', which does not exist
> WARNING: Makefile 'package/feeds/gluon_base/gluon-config-mode-geo-location/Makefile' has a dependency on 'gluon-geolocator', which does not exist
> ```
> 
> Te dependency to `gluon-web-osm` and `gluon-geolocator` should be noted in the readme
> 
> please add an example to the readme, which packages you have to include in your `site.mk`
> 

You are on the wrong branch. This MR was splittend from 
`gluon-config-mode-geo-location` and is in the branch 
`gluon-config-mode-geo-location-with-map`.

But you are right I should mention the dependency on `gluon-web-osm`

cheers
Tarek

rubo77 commented:
But the branch `gluon-config-mode-geo-location-with-map`

doesn't contain the package  `gluon-web-osm`

that's why I created my own branch, where i merged those two

2tata commented:
On 07/21/18 11:51, Ruben Barkow wrote:
> But the branch `gluon-config-mode-geo-location-with-map`
> 
> doesn't contain the package  `gluon-web-osm`

Yes, you need to merge the branch gluon-web-osm from #1483

rubo77 commented:
i did merge it, and now I get this error :
```

Collected errors:
 * check_data_file_clashes: Package gluon-config-mode-geo-location-with-map wants to install file /media/ruben/linuxbackup/gluon/lede/openwrt/build_dir/target-mips_24kc_musl/root-ar71xx$
lib/gluon/config-mode/wizard/0400-geo-location.lua
        But that file is already provided by package  * gluon-config-mode-geo-location
 * opkg_install_cmd: Cannot install package gluon-config-mode-geo-location-with-map.

```

2tata commented:
On 07/21/18 20:23, Ruben Barkow wrote:
> So to use this package, instead of the feature `web-wizzard` I habe to add all packages from that feature manually and replace `gluon-config-mode-geo-location` with  `gluon-config-mode-geo-location-with-map`?
> 
> like this? https://github.com/freifunk-kiel/site-ffki/commit/8881f34778ce3021dc266cc1b6c7bc8c4302cb74
> 

No, like it is explaned in the gluon documentaion:

GLUON_FEATURES := \
	config-mode-mesh-vpn \
	web-private-wifi \
	ebtables-limit-arp \
	ebtables-filter-multicast \
	ebtables-filter-ra-dhcp \
	mesh-batman-adv-14 \
	mesh-vpn-fastd \
	radvd \
	respondd \
	status-page \
	web-advanced \
	web-wizard \
	autoupdater

  GLUON_SITE_PACKAGES += \
	-gluon-config-mode-geo-location \
  	gluon-config-mode-geo-location-with-map
	...

cheers
Tarek

rubo77 commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,57 @@
+  - defaults to ``0.0``
+
+The map will center to the position defined in map_lat and map_lon, unless a
++location was already defined. The new defined location will become the
```
"unless a +location" is not clear, what you mean with that.

maybe just:

"The map will center to the position defined in `map_lat` and `map_lon`, unless a location was already selected by the user before."

delete this line: "The new defined location will become the new center."

rubo77 replied:
Was willlst du denn mit dem satz erklären? 

Hat es etwas mit Cookies zu tun? dass man wenn man das 2. mal auf die Karte kommt er schon die Position des ersten mals auswählt? Oder was meinst du mit "already defined"?

Wenn du das mal kurz auf deutsch erklärst, kriegen wir das schon hin.

2tata replied:
@rubo77 Es geht darum wenn man ein weiteren mal in den configmode geht und zuvor bereits eine position definiert hat, dann wird diese als center genutzt.

rubo77 commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,57 @@
+
+config_mode.geo_location.map_lon \: optional
+  - represents the default longitude center of the location picker map.
+  - defaults to ``0.0``
```
set the default to the center of europe instead of the Golf of Guinea.

mweinelt replied:
Another approach: treat missing value as nil and don't show the map.

2tata commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,55 @@
+==============================
+
+This package allows the user to set latitude, longitude and optionally altitude
+within config mode. There are 2 types of this package.
```
> I suppose it is not possible to merge the two packages and have a site.conf option decide whether to enable the feature, because the site.conf values are not available in the Makefile, only environment variables will be, correct?

I am not sure what do you mean. There are two completly seperated packages where you need to add to the site.mk. Depending on which package you have put into the site.mk you will have optional additional parameters in the site.conf

> The package is split off to prevent cluttering the filesystems with all required js/css files, when the map is not wanted, correct?

Its split off, to give the communities different options. Maybe some communities don't want to have a map in the config-mode. Therefore it is not necessary to have dead code on the Routers flash.

rubo77 commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,57 @@
+
+It is possible to include ``gluon-config-mode-geo-location`` or
+``gluon-config-mode-geo-location-with-map`` in the ``site.mk``.
+
```
It is possible to include **either** ``gluon-config-mode-geo-location`` **or**
``gluon-config-mode-geo-location-with-map`` in the ``site.mk``.

If you want to use `gluon-config-mode-geo-location-with-map` together with the GLUON_FEATURE `web-wizard` then you have to exclude `gluon-config-mode-geo-location`, i.e.:

```
GLUON_SITE_PACKAGES += \
  -gluon-config-mode-geo-location \
  gluon-config-mode-geo-location-with-map
```

2tata replied:
I Think this is redundancy in the gluon doc ... but I'll include it anyway.
applied

rubo77 commented:
I built the package in our firmware, but:

### 1. the translation is not selecting the german sentences.

It could have to do something with a change happened here: https://github.com/freifunk-gluon/gluon/commit/60522ee2532b3e34569fa3bf027ad3aead4d6ed9#diff-ed87e56578bc9c1713d3f040c546ffad

### 2. There is no map!

![auswahl_005](https://user-images.githubusercontent.com/1591563/43040704-faa82a84-8d4a-11e8-9e5d-a024814b95ff.png)

This is the site.mk I used:
https://github.com/freifunk-kiel/site-ffki/blob/geo/site.mk#L60

Adorfer commented:
to make the map and setup work on a computer "still in normal internet": 
I would suggest to move the setup-page to an IPv6-Linklocal-address.  and announce this perhaps with radvd/dnsmasq with name like "setup.gluon", so it will be reachable as http://setup.gluon. 
This will then not interfere with all other uplinks of client computers.

2tata commented:
On 07/22/18 01:05, Ruben Barkow wrote:
>   I built the package in our firmware, but:
> 
> ### 1. the translation is not selecting the german sentences.
> 
> It could have to do something with a change happened here: https://github.com/freifunk-gluon/gluon/commit/60522ee2532b3e34569fa3bf027ad3aead4d6ed9#diff-ed87e56578bc9c1713d3f040c546ffad
> 
> ### 2. There is no map!
> 
> ![auswahl_005](https://user-images.githubusercontent.com/1591563/43040704-faa82a84-8d4a-11e8-9e5d-a024814b95ff.png)
> 
> This is the site.mk I used:
> https://github.com/freifunk-kiel/site-ffki/blob/geo/site.mk#L60
> 

don't use `olurl = 'http://osm.ffnw.de/.static/ol/OpenLayers.js'` this doesn't 
exist anymore and is just an example URL.

The map will only shown if the source loadable. So thus the above example URL 
doesn't exist anymore it is not loadable. Please Use the default `olurl`.

cheers
Tarek

2tata commented:
On 07/22/18 04:10, Adorfer wrote:
> to make the map and setup work on a computer "still in normal internet":
> I would suggest to move the setup-page to an IPv6-Linklocal-address.  and announce this perhaps with radvd/dnsmasq with name like "setup.gluon", so it will be reachable as http://setup.gluon.
> This will then not interfere with all other uplinks of client computers.

Hm, sounds interesting. But what about the case of redirecting from the factory 
Firmware to the gluon site. Not sure how does it works in general. but 
redirecting from e.g. 192.168.0.1 -> 192.168.1.1 should be easier than to 
IPv6-linklocal?

But I think that should be discuss on an another place like an issue or IRC.

cheers
Tarek

2tata commented:
@mweinelt zoom level is now configurable over site.conf.

mweinelt commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -37,29 +37,36 @@ config_mode.geo_location.show_altitude \: optional
+         map = {
+             lon = 52.951947558,
+             lat = 7.844238281,
+             zoom = 12,
```
So now that we have zoom, do we need another zoom level for when a position is selected?

2tata commented:
On 07/22/18 15:31, Martin Weinelt wrote:
> mweinelt commented on this pull request.
> 
> 
> 
>>     - ``url`` set an URL for OpenStreetMap layers.
>     - defaults to ``http://dev.openlayers.org/OpenLayers.js``
>   
>   Example::
>   
> -  config_mode = {
> -            geo_location = {
> -                    map_lon = 52.951947558,
> -                    map_lat = 7.844238281,
> -                    openlayers_js_url = 'http://osm.ffnw.de/.static/ol/OpenLayers.js',
> -                    show_altitude = true,
> -            },
> -  },
> + config_mode = {
> +     geo_location = {
> +         map = {
> +             lon = 52.951947558,
> +             lat = 7.844238281,
> +             zoom = 12,
> 
> So now that we have zoom, do we need another zoom level for when a position is selected?
> 
I think we should keep the number of parameters in the site.conf on a minimum. 
So I would say no we don't need that.

cheers
Tarek

2tata commented:
On 07/22/18 15:33, Martin Weinelt wrote:
> Another approach: treat missing value as nil and don't show the map.

nil is use for the default vailue 0.0.

mweinelt commented:
At least the zoom level should be higher when a position was selected. This encourages repicking a better position.

2tata commented:
> At least the zoom level should be higher when a position was selected. This encourages repicking a better position.

The users can still zoom into the map by scrolling or by using the scale bar on the site.

rubo77 commented:
I get the error

    Laden fehlgeschlagen für das <script> mit der Quelle "http://dev.openlayers.org/OpenLayers.js%60?noCache=1523513258".

%60 ist ein backtick: ```

2tata commented:
On 07/23/18 13:46, Ruben Barkow wrote:
> I get the error
> 
>      Laden fehlgeschlagen für das <script> mit der Quelle "http://dev.openlayers.org/OpenLayers.js%60?noCache=1523513258".
> 
> %60 ist ein backtick: ```

Hm, that musst be somethin from you. There isn't any backtick in the code Maybe 
you put something in the site.conf ?

cheers
Tarek

rubo77 commented:
I did, as i posted above. removing the backtick in `lib/gluon/site.json` makes it work:

![auswahl_010](https://user-images.githubusercontent.com/1591563/43075236-767a7fdc-8e80-11e8-95e7-e0a025b2b44f.png)

Now, I think the text "Um deinen Knoten auf der Karte anzeigen zu können, ..." should be below the map, not above and there should be 10px more padding from the map to the text

rubo77 commented:
and if you forget to enter a hostname and click the save button, the map doesn't keep its zoomlevel on the reloaded page

2tata commented:
> and if you forget to enter a hostname and click the save button, the map doesn't keep its zoomlevel on the reloaded page
@mweinelt May I should store the last use zoom level as UCI parameter? Do you mean something like that?

rubo77 commented:
Not UCI parameter but maybe in the cookie or in a hidden input field, so you have it on a right load of the page.

If you leave the page and then come back later it does not need it to have a pre selected zoom level from the last time

rubo77 commented:
Or even better just use the coordinates that are stored in the input Fields for longitude and latitude to automatically zoom the map there if they are present

2tata commented:
On 07/23/18 15:43, Ruben Barkow wrote:
> Or even better just use the coordinates that are stored in the input Fields for longitude and latitude to automatically zoom the map there if they are present
> 

I don't know what du you mean here. But regarding to the reset of the zoom level 
I don't know how to solve that.

cheers
Tarek

2tata commented:
On 07/23/18 13:59, Ruben Barkow wrote:
> I did, as i posted above. removing the backtick in `lib/gluon/site.json` makes it work:
> 
> ![auswahl_010](https://user-images.githubusercontent.com/1591563/43075236-767a7fdc-8e80-11e8-95e7-e0a025b2b44f.png)
> 
> Now, I think the text "Um deinen Knoten auf der Karte anzeigen zu können, ..." should be below the map, not above and there should be 10px more padding from the map to the text
> 

I prefer the text above the map.

I don't know how to fix that padding. I add some padding-top but it doesn't react.

rubo77 commented on docs/package/gluon-config-mode-geo-location.rst:
```diff
@@ -0,0 +1,72 @@
+Example::
+
+ config_mode = {
+     geo_location = {
```
indention should be 2 spaces (like everywhere in gluon)

rubo77 commented:
Another issue: I am not lucky with the dropdown-box with only two options. What exactly is the effect of "Disabled" and of "static location"?

I guess, that you could also create a "dynamic location" if you add some extra module.

so if there are only those two options they are obsolete by the checkbox "share your location on the map" which seems the same to the user.

Maybe best to automatically select "static location" and make the dropdown invisible if there is no third option.

2tata commented:
On 07/23/18 22:38, Ruben Barkow wrote:
> rubo77 commented on this pull request.
> 
> 
> 
>> +
> +define Package/gluon-config-mode-geo-location-with-map/install
> +	$(CP) $(PKG_BUILD_DIR)/gluon-config-mode-geo-location-with-map/luadest/* $(1)/
> +	$(INSTALL_DIR) $(1)/lib/gluon/config-mode/www/static/
> +	$(LN) /lib/gluon/web/www/static/osm.js $(1)/lib/gluon/config-mode/www/static/osm.js
> +	$(call GluonInstallI18N,gluon-config-mode-geo-location-with-map,$(1))
> +endef
> +
> +define Package/gluon-config-mode-geo-location/postinst
> +#!/bin/sh
> +$(call GluonCheckSite,$(PKG_BUILD_DIR)/gluon-config-mode-geo-location/check_site.lua)
> +endef
> +
> +define Package/gluon-config-mode-geo-location-with-map/postinst
> +#!/bin/sh
> +$(call GluonCheckSite,$(PKG_BUILD_DIR)/gluon-config-mode-geo-location-with-map/check_site.lua)
> 
> this is obsolete then

No, its not, the gluon.mk is to limited, so I defined my own call defs.

2tata commented:
On 07/23/18 23:29, Ruben Barkow wrote:
> Another issue: I am not lucky with the dropdown-box with only two options. What exactly is the effect of "Disabled" and of "static location"?

The Disable state erase all location parameters. For People who do not want any 
locatable information on Routers.

> I guess, that you could also create a "dynamic location" if you add some extra module.

The dropdown-box was inserted for various options of integrating Positions, 
where some more will follow in future.

> so if there are only those two options they are obsolete by the checkbox "share your location on the map" which seems the same to the user.
> 
> Maybe best to automatically select "static location" and make the dropdown invisible if there is no third option.

The "share your location on the map" checkbox is indepentent from the location 
it's self, because maybe people would like to define a position but not want to 
share it.

cheers
Tarek

rubo77 commented:
Then only show both options, if it makes sense to add coordinates and not show them.
without a proper package that uses entered-not-shared coordinates it makes no sense to the user to have those two options.

So: Hide one of them if the extended package is not installed.


---
about the CSS: you can add `!important` to the map layer to get a margin:

```
#locationPickerMap {
    margin-top: 20px !important;
    margin-bottom: 20px !important;
}
```

but maybe that wouldn't be deeded, if you move the may layer out of the div for the infotext.

rubo77 commented:
> this is obsolete then

i was just talking about the review vomment, which is obsolete now, because it can stay as it was (without indention)

You can delete your (very large) comment https://github.com/freifunk-gluon/gluon/pull/1484#issuecomment-407472141, then I delete this one

2tata commented:
On 07/25/18 12:25, Ruben Barkow wrote:
>>> On 07/23/18 15:43, Ruben Barkow wrote:
>>>   Or even better just use the coordinates that are stored in the input Fields for longitude and latitude to automatically zoom the map there if they are present
>>
>> I don't know what du you mean here. But regarding to the reset of the zoom level
>> I don't know how to solve that.
> 
> I was mistaken, the map is already located at exactly the coordinates, stored in the node. so everything is fine there.
> 
> Now at the moment, the map is quite zoomed out when you open the config mode and already have entered coordinates. the map should be zoomed in maximum in that case, so you see, where your pin is located on the map exactly.

Now, I know what do you mean, I have fix it now. If there coordinates present. 
The map will zoomed as max as can.

2tata commented:
Now the map is loading asyncron from the rest of the formula. Means the page loading is not blocked by external sources anymore, because all external sources will by loaded on java script runtime.
@rotanid @mweinelt @rubo77 for me the MR is done now.

2tata commented:
On 07/28/18 13:02, Ruben Barkow wrote:
> rubo77 commented on this pull request.
> 
> Maybe we should maximal minify this by using only one-character variables?

I think the benefit of minifying this small part is outweigh by the readability 
of the code.

cheers
Tarek

xf- commented:
License is missing for OSM https://www.openstreetmap.org/copyright

rotanid commented on package/gluon-config-mode-geo-location/Makefile:
```diff
@@ -6,8 +6,82 @@ PKG_VERSION:=1
+	$(call GluonInstallI18N,$(1),$(PKG_BUILD_DIR)/gluon-config-mode-geo-location-with-map)
+endef
+
+define Package/gluon-config-mode-geo-location/postinst
```
some parts of your Makefile could be obsolete because of the treewide changes to package makefiles which were made in March&April.
please check this yourself.

2tata replied:
I have already check that. It is necessary to create thous defines in the Makefile because of the changing directories.

rubo77 commented:
If you select "manuelle Position" the hook in "Position für die Karte freigeben" should automatically be selected, otherwise, you could forget to set the hook easily

rubo77 commented:
Also "Manuelle Position" doesnt make any sense, it there is no "automatic position" or such.

this should be renamed into "select position on map" instead

rubo77 commented on package/gluon-config-mode-geo-location/src/de.po.in:
```diff
@@ -0,0 +1,53 @@
+msgstr "Geo-Position"
+\n
+msgid "Static location"
+msgstr "Manuelle Position"
```
Also "Manuelle Position" doesnt make any sense, if there is no "automatic position" or such as another option.

this should be renamed into "select position on map" instead

2tata replied:
Hm, the user selects manual a position on the map or by hand. So this should by okay for me.

rubo77 replied:
"manual position" translates to "Handstellung" :)

even if not: as I said, the word "manual" implements that there is a "not-manual" way also, but tif there isn't it is just strange to emphasize "manual".

I think "select position on map" suits better to all cases

2tata commented:
> License is missing for OSM https://www.openstreetmap.org/copyright

Fixed in 1b07343

> No option for own tiles

What do you mean? You have the option to set your own URL for your layers.
 
> (No SSL for in default JS (Chrome 68 warning Not Secure - don't know if private IP has an exception)

Fixed in 1b07343 . All external URLs are now with SSL.

> JS and tiles are hosted in US -> need to add/adjust privacy policy

How, do you know that?

xf- commented:
>  What do you mean? You have the option to set your own URL for your layers.
How to provide an option for own tiles/layers

> How, do you know that?
I took a look in your default JS - loaded the map and took a look at inspect tool. Is a project from a US university.

rubo77 commented:
Mal auf Deutsch:

Irgendwie ist das immer noch nicht so richtig logisch. "Disabled" Ist auch Quatsch, wenn man dann trotzdem sagen kann: "Share your location".

- Was passiert denn genau bei den beiden Optionen jeweils?
- Welche 3. Option ist geplant?
- Was ist die Alternative zur Karte?


Vielleicht finden wir noch eine Formulierung, die auf alles passt.

2tata commented:
On 08/07/18 01:13, Ruben Barkow wrote:
> 
> Mal auf Deutsch:
> 
> Irgendwie ist das immer noch nicht so richtig logisch. "Disabled" Ist auch Quatsch, wenn man dann trotzdem sagen kann: "Share your location".

Bei "Disabled" kann man "Share your location" auch nicht mehr auswählen und es 
wird auch nicht angezeigt.

> - Was passiert denn genau bei den beiden Optionen jeweils?

"Static Position" Zeigt lat/lon/altitude Felder zur manuellen Eingabe von 
Koordinaten sowie die Option das diese veröffentlicht werden, z.b. für eine map.
als Feature könnte eine Map sichtbar sein die dem User hilft eine Position zu 
setzen.

"Disabled" Deaktiviert und Löscht (falls vorhanden) die Hinterlegen 
Informationen. (Sprich keine Angabe von Koordinaten).

> - Welche 3. Option ist geplant?

3. wären "Dynamic Position" der geolocator mit rein dynamischer Position, sowie 
4. "Dynamic & Static Position" Dies erlaubt das feste setzen einer Position 
gleichzeitig aber auch das scannen und übertragen von umliegenden WLANs an das 
z.Z. noch openwifi.su backend.

5. Mögliche Variante wäre die Berechnung der eigenen Position aus über respondd 
mitgeteilten Koordinaten umliegender direkter Nachbarn, also eine Art cooplocator.

Die o.g. Möglichkeiten sind nur Ideen und sind für den MR nicht vorgesehen 
sondern sollen künftig in ähnlicher Form folgen.

> - Was ist die Alternative zur Karte?

Position über die Tastatur eintippen?

cheers
Tarek

rubo77 commented:
Ok, da das Label ja schon "Geo-Position"ist, einfach

- deaktivieren
- anzeigen
- automatisch bestimmen

Die 3. Option kann man so später in einem anderen Modul hinzufügen

rubo77 commented:
If there is no 3rd module, the whole first option must be hidden completely, or what sense is there to add coordinates into the node if you don't want to share it?

rotanid commented:
@rubo77 you may add these coordinates so people with remote access to the node can check its location. so there is a use case for adding coordinates without sharing them.

rubo77 commented:
That sounds not like a use case (who would want that and why? if you want to give the coordinates to womeone with root access to your node, you could simply put a text file on the console).

We should keep the config mode main page really simple, that is the main strength of Gluon compared to other firmwares.

So one checkbox to enter coordinates is enough.

If there later is an extra package, that makes sense to enter coordinates but keep them secret, then we should add the suggested extra checkbox, but only then.

rotanid commented:
@rubo77 sure, this PR doesn't have to introduce this if it isn't possible at the moment. i just wanted to add that you and me not having the use case doesn't mean there isn't any.

NeoRaider commented:
I've only had a short look at the code so far, but I don't like using the preprocessor to conditionalize code at all. I would like to clean this up a bit before it is merged (I might take care of it this weekend).

NeoRaider commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,148 @@
+					uci:set("gluon-node-info", location, "altitude", altitude.data)
+				else
+					uci:delete("gluon-node-info", location, "altitude")
+				end
```
This is cargo-culting from luci.model.uci. With simple-uci, `uci:set(p, s, o, nil)` is equivalent to `uci:delete(p, s, o)`, so you usually don't need to check where a value is set before passing it to uci:set.

NeoRaider commented on package/gluon-config-mode-geo-location/src/0400-geo-location.lua.in:
```diff
@@ -0,0 +1,148 @@
+			uci:delete("gluon-node-info", location, "altitude")
+			uci:delete("gluon-node-info", location, "latitude")
+			uci:delete("gluon-node-info", location, "longitude")
+			uci:set("gluon-node-info", location, "share_location", 0)
```
Don't use numbers where a boolean is semanctially saner (this is another thing you could not do with luci.model.uci).

NeoRaider commented:
I've finished my cleanup work: #1515
